### ISSUE TITLE
BotConnectorBot and TextBot behaves differently with dialogs

### ISSUE THREAD

**Issue by Aleksion** [Microsoft/BotBuilder#26]
Since I'm sitting on a mac, the way I debug my code is by runningt a textbot locally. To test everything out, I'm building a simple conversational bot, which goes through a simple waterfall model. 

I have a three step waterfall model which can exit at multiple stages. The way I do this is by calling session.endDialog() if the exit criteria is met. If it isn't met, I call session.send('Current state') and then I present a new prompt (builder.Prompts.choice(session, "Do you want me to?)). 

This works just fine locally - however, I can't seem to get it to work as intended when it's online. To me it seems like the botConnectorBot exists the dialog when send() is called.  Is this correct? And if that is the case - how would I go about presenting the user with intermediary data? (I could of course just include it in the prompt, but they can get quite hefty then).

Furthermore, when I couple my botConnectorbot with slack it behaves weirdly. It triggers the one match I've enabled even though there is no matching text: 

### Text bot
![screen shot 2016-04-01 at 12 21 59](https://cloud.githubusercontent.com/assets/3367641/14204441/ac7202f0-f804-11e5-8466-eb47d479174c.png)

### Bot Connector Bot through slack
<img width="513" alt="screen shot 2016-04-01 at 12 23 21" src="https://cloud.githubusercontent.com/assets/3367641/14204452/be9cfaa2-f804-11e5-9b0d-bf52b63482b3.png">

**Comment by Stevenic** [comment]
You should be able to call session.send() multiple times but to be honest the TextBot is currently the most tested path as that's what the Unit Tests use. Would it be possible for you to share the code with your waterfalls? Feel free to replace the business logic with a comment like "// doing something here". I just want to see how your dialogs hang together and I'll see if I can replicate the issues you're seeing.  If I can replicate it I can fix it :)

**Comment by Stevenic** [comment]
And by the way... calling session.endDialog() from any step in the waterfall should be fine. That was how I intended things to work.

**Comment by Stevenic** [comment]
Oh... one other question. Which version of the npm module are you using? Make sure you're at least running 0.6.1 as we did push a fix yesterday that updates the URL we send out of band messages to for BotConnectorBot (your secondary message.) The latest version is 0.6.2 and I'm about to have to push a 0.6.3

**Comment by tomlm** [comment]
Make sure your botdata doesn't have old data from previous bots...As people change their schema they aren't clearing out the old data.  Since the emulator never persists the bot data devs aren't running into this on their box, but only when they publish it into a place where they had an older bot with older bot data.

**Comment by Aleksion** [comment]
Thx for all the responses! 
I will get back to you @Stevenic! 

@tomlm  - where should the botdata be stored and how can I clear it?

**Comment by Stevenic** [comment]
@Aleksion BotBuilder hides all of this from you. It does a basic sanity check over the BotData that's coming in and if it thinks you've made radical changes to things it will clear it for you but Tom is correct there are a couple of edge cases where you could be getting something in with bad data. I don't think that's the case based on what you've described but if you want to make sure. Temporarily modify your bot to have a single root '/' dialog that does the following:

    bot.add('/', function (session) {
        session.endDialog();
    });

This will ensure that everything is reset.

**Comment by Aleksion** [comment]
Hi @Stevenic  - here's my waterfall code. It's a bit hacky - my goal was to try to get a functional bot as soon as possible :) I've removed some of the code for clarity (mainly in the requests)

```
session => {
    builder.Prompts.text(session, 'What is the name of the company you want to search for?')
  },
  (session, results) => {
    session.userData.companyName = results.response;

    request(REQUEST_OBJECT, (error, response, body) => {
      if (!error && response.statusCode == 200) {
        //console.log(body.company)
        session.userData.company = JSON.parse(body);
        session.send(`The first company I found was ${session.userData.company.name} with VAT:${session.userData.company.vat}`);

        builder.Prompts.choice(session, "Do you want me to fetch their accounts?", "Yes|No");
      } else {
        session.send("I'm sorry - couldn't find anything :(");
        session.endDialog();
      }
    })
  },
  (session, results) => {
    if (results.response && results.response.entity === "Yes") {
      var vat = session.userData.company.vat;
      request(SECOND_REQUEST_OBJECT, (error, response, body) => {
        if (!error && response.statusCode == 200) {
          var parsedResponse = JSON.parse(body);
            
            session.send(`You can find the latest accounts published on ${parsedResponse.url}`);
            session.endDialog();
          }else {
            session.send(`I couldn't find any documents for ${session.userData.company.name}`);
            sesison.endDialog();
          }


        } else {
          session.send("I'm sorry - couldn't find anything :(");
          session.endDialog();

        }
      })
    } else {
      session.send("Glad to be of service!");
      session.endDialog();
    }
  }
```

I wonder if it's my async calls that messes things up? (though it does work flawlessly with the TextBot

Let me know if you need any clarification

**Comment by Stevenic** [comment]
@Aleksion Lets break this into pieces. First you're definitely running into issues with sending back-to-back messages. We haven't tested that very thoroughly so I'll have to look into what's happening. We may have to introduce a small delay between messages to keep the switch from getting confused. For now I'd merge your send statement into your prompt so.

    builder.Prompts.confirm(`The first company I found was ${session.userData.company.name}. Do you want me to fetch their accounts?`);

I also switched to using Prompts.confirm() which had a bug that I fixed in the latest build.  This should address that issue.

The second issue you're seeing is that it's dumping straight into the waterfall when you say "Hi!". My question there is are you using a CommandDialog or something? What are you passing your waterfall to? I'd need to see that code.

**Comment by Aleksion** [comment]
When I wrote the code, confirm didn't work as you just mentioned. I'd much rather use that one, so thank you for the fix!

I'm using a CommandDialog as such:

```
new builder.CommandDialog()
  .matches('^cvr', builder.DialogAction.beginDialog('/cvr'))
  .onDefault(session => {
    if (session.userData.company) {
      session.send('This is the company you last searched for: %s, with VAT: %s ', session.userData.company.name, session.userData.company.vat);
      //session.send('With the VAT:', session.userData.company.vat);
    } else  {
      session.send('If you want to start a search, type: cvr');
    }
  })
```

**Comment by Aleksion** [comment]
Add the message did as expected fix the first issue. I'll stick to that until #44 is fixed

**Comment by Stevenic** [comment]
Definitely acts like its bypassing the CommandDialog and going straight to the '/cvr' dialog. So from this I'm assuming CommandDialog is added as the root dialog '/' and then you have your waterfall added as '/cvr'. Also you didn't change the default right?

**Comment by Stevenic** [comment]
So this is going to sound crazy but it looks like ^cvr is matching when you say 'Hi!' could you replace the beginDialog action with:

    session => { session.send('I heard: ' + session.message.text); }

**Comment by Aleksion** [comment]
Yep - the command dialog is added at the root.

I've broken the application up a bit. I'll add the structuring code for clarity:

##### registerDialogs.js:
```
var cvrDialog = require('./cvr');
var indexDialog = require('./');

module.exports = function registerDialogs(bot) {
  bot.add('/', indexDialog);
  bot.add('/cvr', cvrDialog);
}
```

##### index.js
```

var builder = require('botbuilder');
var cvr = require('./cvr');

module.exports = new builder.CommandDialog()
  .matches('^cvr', builder.DialogAction.beginDialog('/cvr'))
  .onDefault(session => {
    if (session.userData.company) {
      session.send('This is the company you last searched for: %s, with VAT: %s ', session.userData.company.name, session.userData.company.vat);
      //session.send('With the VAT:', session.userData.company.vat);
    } else  {
      session.send('If you want to start a search, type: cvr');
    }
  })
```

**Comment by Stevenic** [comment]
my suspicion is we're getting something like cvrbot: Hi!

**Comment by Aleksion** [comment]
Nope - it acts as suspected:
<img width="513" alt="screen shot 2016-04-01 at 12 23 21" src="https://cloud.githubusercontent.com/assets/3367641/14233747/94e41bfe-f9d1-11e5-8e76-b37700dbc480.png">

**Comment by Aleksion** [comment]
I've narrowed down the problem. It seems like it never returns to the root after beginning the dialog. (I removed the else statement in the onDefault to simplify things)
 
<img width="528" alt="screen shot 2016-04-03 at 19 29 13" src="https://cloud.githubusercontent.com/assets/3367641/14233789/625fcf74-f9d2-11e5-8632-f8731123d654.png">

**Comment by Stevenic** [comment]
Now the dots are connecting for me...

Yes... We definitely have a problem with back-to-back calls to the server. This may not make a lot of sense but when you're calling session.send() followed by session.endDialog() we're triggering 2 calls to the server. One to send the message and a second to say that the dialog has ended. For now here's the fix...

When you want to end your dialog call:

    session.endDialog({ response: 'message to send user' });

Then on the CommandDialog side do this:

    .matches('^cvr', (session, results) => {
        if (results && results.response) {
            session.send(results.response);
        } else {
            session.beginDialog('/cvr');
        }
    }) 
 
This should resolve your issue for now because it lets the send() persist the fact that the dialog has ended. I'm going to update the DialogActions.beginDialog() call to support the same basic behavior. You can end the dialog with a message to send to the user.

**Comment by Aleksion** [comment]
Sounds like an event driven architecture for your backend would make sense - so you ensure that the final state is always a result of the events, in the order that they were received.

And then the botbuilder framework would also need to be implemented as such, so that a race condition between session.send() and session.endDialog() can't occur.

Thx for responding so swiftly @Stevenic ! Do you guys ever take the weekend off? ;) (you can close it unless you want to track more with this issue)

**Comment by Aleksion** [comment]
Just pushed and confirmed. That did fix the issue.

**Comment by tvmick** [comment]
I've been trying to unsubscribe from this thread and indeed this site for over a wk , GitHub not answering my pleading for info to unsubscribe and the people on this thread are using it like an old CB radio and talking all night stopping me from sleeping and at 76 yrs old it's not Good , would you like your grandad kept awake all night and I derntswitch mobile of it my only life line

**Comment by Aleksion** [comment]
@tvmick - unwatch the repository?

And repositories like this will have "chatter all night", since it's being used globally.

**Comment by Stevenic** [comment]
@tvmick yes - you should unwatch. This is a new library so we have a lot of issues we need to work through. We're not trying to spam you but at the same time we're committed to helping our users solve their problems. Sorry.

**Comment by Stevenic** [comment]
@Aleksion I just checked in a change that will be in the next release which lets you send a message as part of calling endDialog(). This should let you go back to using the DialogActions.beginDialog() as you were.  I'm not going to push the change to NPM yet as there's a valid work around and I'd like to accumulate some more changes.

**Comment by Stevenic** [comment]
I just published 0.7.0 which adds the ability to send a message when calling endDialog()

**Comment by mikkhait** [comment]
Sort of related to the same topic. I'm using 0.8.0 and trying to use `session.send()` site-by-side...

```
session.send("hey, good to see you here");
session.send("let's being");
```
while this works in _TextBot()_, when using with _BotConnector()_ session ends upon first `session.send()`
